File: codigo/Web/cosas/backend/WebScraping/WebScraping.php
<?php
include_once('simplehtmldom_1_9_1/simple_html_dom.php');

function get_amazon2($product_id){
    $url_amazon = "https://www.amazon.es/";
    $url_buscar = $url_amazon . "s?k=" . $product_id;
    $html = file_get_html('https://www.amazon.es/s?k=6972453163820');
    $link = $html->find('a.a-link-normal', 0)->href;
    return $link;
        
        
        
        
        
        
        
        $url_product = $html->find("h2 > a", 0)->href;
        
        $url_final = $url_amazon . $url_product;
        return $url_final;
    return $urlProduct;
    if ($urlProduct != "") {
        $data = get_product_data_amazon($urlProduct);
        return $data;
    } else {
        return array("product_name", "img", "descriptor", "specs", "price");
    }
}


function get_product_data_amazon($url){
    try {
        $html = get_html($url);
        $product_name = $html->find("#productTitle", 0)->plaintext;
        $img = $html->find("#imgTagWrapperId img", 0)->src;
        $descriptor = $html->find("#feature-bullets > ul > li", 0)->plaintext;
        $decimal = explode(",", $html->find(".a-price-whole", 0)->plaintext)[0];
        $fraction = explode(" ", $html->find(".a-price-fraction", 0)->plaintext)[0];
        $price = floatval($decimal . "." . $fraction);
        $specs = "";
        try {
            $table = $html->find("#productDetails_techSpec_section_1", 0);
            $rows = $table->find("tr");
            foreach ($rows as $row) {
                $cells = $row->find("td, th");
                $attribute = $cells[0]->plaintext;
                $value = $cells[1]->plaintext;
                $specs .= $attribute . ": " . $value . "\n";
            }
        } catch (Exception $e) {
            $specs = "The table does not exist on the page";
        }
        return array($product_name, $img, $descriptor, $specs, $price);
    } catch (Exception$e) {
        echo$e;
        return array("product_name", "img", "descriptor", "specs", "price");
    }
}
function get_hola(){
    return array("product_name", "img", "descriptor", "specs", "price");
}


include_once('simplehtmldom_1_9_1/simple_html_dom.php');

function get_amazon($url){
    $html = file_get_contents($url);
    $doc = new DOMDocument();
    @$doc->loadHTML($html);
    $xpath = new DOMXPath($doc);
    $links = $xpath->query('//*[@id="search"]/div[1]/div[1]/div/span[1]/div[1]/div[3]/div/div/div/div/div[2]/div[1]/h2/a');
    if ($links->length > 0) {
        $href = $links[0]->getAttribute('href');
        return $href;
    }
    return null;
}

$product_id = $argv[2];
$url_amazon = "https://www.amazon.es/";
$url_buscar = $url_amazon . "s?k=" . $product_id;
echo "Url buscar: ".$product_id;
$link = get_amazon($url_buscar);
echo "\nUrl producto:".$link;
?>

File: codigo/Web/cosas/backend/WebScraping/WebScraping.py
import json
import sys
from bs4 import BeautifulSoup
import requests

def get_amazon(product_id):
    urlProduct = get_product_amazon(product_id)
    if urlProduct != "":
        data = get_product_data_amazon(urlProduct)
        return data
    else:
        return ["product_name", "img", "descriptor", "specs", "price"]

def get_product_amazon(product_id):
    url_amazon = "https://www.amazon.es/"
    url_search = url_amazon + "s?k=" + product_id
    try:
        response = requests.get(url_search)
        soup = BeautifulSoup(response.text, 'html.parser')
        url_product = soup.select_one("h2 > a")['href']
        return url_amazon + url_product
    except Exception as e:
        print(e)
        return ""


def get_product_data_amazon(url):
    try:
        response = requests.get(url)
        soup = BeautifulSoup(response.text, 'html.parser')

        product_name = soup.select_one("#productTitle").text
        img = soup.select_one("#imgTagWrapperId img")['src']
        descriptor = soup.select_one("#feature-bullets > ul > li").text
        decimal = soup.select_one(".a-price-whole").text.split(",")[0]
        fraction = soup.select_one(".a-price-fraction").text.split(" ")[0]
        price = float(decimal + "." + fraction)
        specs = ""
        try:
            table = soup.select_one("#productDetails_techSpec_section_1")
            rows = table.select("tr")
            for row in rows:
                cells = row.select("td, th")
                attribute = cells[0].text
                value = cells[1].text
                specs += attribute + ": " + value + "\n"
        except Exception as e:
            specs = "The table does not exist on the page"
        return [product_name, img, descriptor, specs, price]
    except Exception as e:
        print(e)
        return ["product_name", "img", "descriptor", "specs", "price"]


if __name__ == '__main__':
    # Get the function name from the command line argument
    function_name = sys.argv[1]
    # Get the product ID from the command line argument
    product_id = sys.argv[2]
    # Call the function with the product ID as a parameter
    result = globals()[function_name](product_id)
    # Convert the result to a JSON string and print it so it can be captured by the Laravel controller method
    print(json.dumps(result))
